### ISSUE TITLE
different passing types of arguments

### ISSUE THREAD

**Issue by liuyadong** [RcppCore/RcppArmadillo#253]
It seems `matrix` is passed by value, while `array` is passed by reference.
``` r
Rcpp::sourceCpp(code='
  #include <RcppArmadillo.h>
  using namespace arma;

  // [[Rcpp::depends(RcppArmadillo)]]

  // [[Rcpp::export]]
  mat fm(mat x){
    x += 1;
    return x;
  }

  // [[Rcpp::export]]
  cube fa(cube x){
    x += 1;
    return x;
  }'
)

m <- matrix(1, nrow=2, ncol=2)
fm(m)
#>      [,1] [,2]
#> [1,]    2    2
#> [2,]    2    2
m
#>      [,1] [,2]
#> [1,]    1    1
#> [2,]    1    1

a <- array(1, c(2,2,2))
fa(a)
#> , , 1
#> 
#>      [,1] [,2]
#> [1,]    2    2
#> [2,]    2    2
#> 
#> , , 2
#> 
#>      [,1] [,2]
#> [1,]    2    2
#> [2,]    2    2
a
#> , , 1
#> 
#>      [,1] [,2]
#> [1,]    2    2
#> [2,]    2    2
#> 
#> , , 2
#> 
#>      [,1] [,2]
#> [1,]    2    2
#> [2,]    2    2
```

<sup>Created on 2019-03-26 by the [reprex package](https://reprex.tidyverse.org) (v0.2.1)</sup>

**Comment by eddelbuettel** [comment]
I think that is simply a manifestation of the [Rcpp FAQ Question 5.1 ](https://cloud.r-project.org/web/packages/Rcpp/vignettes/Rcpp-FAQ.pdf): a _matrix_ is passed through as is as we efficiently re-use the memory.  So the incoming object is changed.  A _cube_ in Armadillo is not the same as an array in R so _new memory is allocated creating a different object_ which is why the second function does not alter its incoming object.

There are numerous simpler examples out there doing this with _NumericVector_ and then passing an _integer_ vector: same thing, needs a cast, is copied from 32 bit int to 64 bit double and not changed back.

A little confusing at first but it's always been like that.

**Comment by liuyadong** [comment]
Thanks for your explanation. In that case the matrix should be changed while the array should not, but the results are the opposite.

**Comment by eddelbuettel** [comment]
You are confusing _input_ arguments, and return arguments.

```r
R> matin <- matrix(1, 2, 2)
R> matout <- fm(matin)
R> all.equal(matin, matout)
[1] "Mean relative difference: 1"
R> arrin <- array(1, c(2,2,2))
R> arrout <- fa(arrin)
R> all.equal(arrin, arrout)
[1] TRUE
R> 
```

Correct.  The logic is the same though  :).  So apparently `cube` reuses the array, hence not const and input altered -- which I recalled the wrong way above.  For matrix and vector we added 'cleaner' interfaces at some point.   I think Rcpp's NumericMatrix still behaves the other way.

In short, make sure you don't assign to incoming object when you want immutable behaviour.
